### DISCUSSION TITLE
Filtering Search Results by Availability in Milvus Using an External Bitset

### DISCUSSION THREAD

**Discussion by kiran12-lgtm** [milvus-io/milvus#33977] [Q&A and General discussion]
I am currently working with the following setup:

Milvus version 2.3.7, pymilvus version 2.3.6.
A database in Milvus containing 4 million 768-dimensional vectors.
My challenge involves performing a vector search across a large number of IDs, ranging from 10,000 to 500,000. For instance, if I have a query that matches 100,000 documents but only 70,000 of those are available in my inventory. I need to filter out the available items based on this information which I have in the form of a bitset.

The current workflow is as follows:

Query Milvus: Execute a query within Milvus to retrieve document IDs for matched vectors without availability consideration.

Post-Processing: Implement post-filtering using the bitset to isolate only those documents with their corresponding availability bit set to '1'. This results in an O(n) computational complexity, where n is the size of the matched documents.

I am seeking a method or strategy that either reduces this operation to O(1) time complexity or allows me to directly fetch available items via Milvus without needing any post-filtering.

Milvus does support filtering with bitsets (applying filters prior to running the actual approximate nearest neighbor (ANN) search). As I already have an availability bitset on hand, is there any way to utilize this within Milvus?

I also thought of other approach where i have updated my availability in form of arrays in documents like this 

`Items : 

[
{
    "product_id": "12345",
    "product_title-vector": [0.23, 0.45, 0.87, ....]  #this is the vector field of required dimension,
    "store_availability": [0, 2, 14, 3, 5, ...] #upto 2000 Stores
},
{
    "product_id": "12346",
    "product_title-vector": [0.23, 0.45, 0.87, ....]  #this is the vector field of required dimension,
    "store_availability": [0,  5, ...] #upto 2000 Stores
},
{
    "product_id": "12347",
    "product_title-vector": [0.23, 0.45, 0.87, ....]  #this is the vector field of required dimension,
    "store_availability": [0, 5, ...] #upto 2000 Stores
},
{
    "product_id": "12348",
    "product_title-vector": [0.23, 0.45, 0.87, ....]  #this is the vector field of required dimension,
    "store_availability": [0, 3, 5, ...] #upto 2000 Stores
},
{
    "product_id": "12349",
    "product_title-vector": [0.23, 0.45, 0.87, ....]  #this is the vector field of required dimension,
    "store_availability": [0, 2, 14, 3, 5, ...] #upto 2000 Stores
},
]
`
To search available items in relevant stores i was trying to use boolean expressions 

bool_expr = "(store_availability in [0] && store_availability[0]==4) && " + \
            "(store_availability in [1] && store_availability[1]==2) && " + \
            "(store_availability in [2] && store_availability[2]==14)"

search_results = collection.search(
    data=query_vector_resized,
    anns_field="product_title_vector",
    param=search_params,
    limit=10,
    expr=bool_expr,
    output_fields=['product_id', 'product_name', 'store_availability']
)

However the major challenge is inventory updates every 10 Minutes and i can not do so many bulk updates so often
I appreciate any guidance or suggestions on how this can be achieved efficiently.

**Comment by xiaofan-luan** [comment]
what is the reason we need to filtering on so much IDs? even on olap database, it is not common to filter on more than 10000 IDs.

if this a expression from other system, could we use that expression to filter milvus?

**Reply by kiran12-lgtm** [reply]
Imagine it like an e commerce use-case where you have a million items in your inventory, and 3 physical stores to keep availability of items.
Store_1_Bitset : 010101010....... 1 Million Index -> IOWA
Store_2_Bitset : 010101010....... 1 Million Index -> Wisconsin
Store_3_Bitset : 010101010....... 1 Million Index  -> DC


Imagine user search's from DC location, we only want to show available items in DC Query is pretty generic we match close to 160000 documents but only 10000 are available at DC store due to bad weather, My ask is how can i integreate this BITSET info inside milvus to only return available item on that particular store .

I am able to do this postfiltering in solr via bitset using extensions, Can there be any similar solution possible in milvus is what my ask ?

**Comment by xiaofan-luan** [comment]
and other way is to put everything in an array like store_availability

and use store_availability contains([1,2,3,4]) as expression?

**Reply by kiran12-lgtm** [reply]
I just updated my question i thought about that, problem is frequent updates

**Comment by xiaofan-luan** [comment]
each entity is a good, and filter is based on where you can buy it? some times the store is out of stock, so you need update store availability?

**Reply by kiran12-lgtm** [reply]
Yes Store availability updates every 15 Minutes.

**Reply by xiaofan-luan** [reply]
that might be too much for milvus.
what is the latency of QPS requirment?

**Reply by kiran12-lgtm** [reply]
100 TPS on a multi query node setup with P95 ~ 250 ms

**Comment by xiaofan-luan** [comment]
we have a lucene like index,  but update 4m data in every 15 minutes seems to be challeging

**Reply by xiaofan-luan** [reply]
maybe you can try to see how the performance looks like. even for solr update that much data might not be very performant.

or you can bulkinsert to another collection, rebuild index and serving with the new collection?

**Reply by xiaofan-luan** [reply]
tantivy index should be able to handle the filter efficient.

@zhagnlu might be a good use case for bitset index

**Reply by kiran12-lgtm** [reply]
> maybe you can try to see how the performance looks like. even for solr update that much data might not be very performant.
> 
> or you can bulkinsert to another collection, rebuild index and serving with the new collection?

We Do not update our catalog(indexed in solr)  we just update our bitsets(which stores inventory), Within solr it provides you ability to have custom components which can do filtering via bitsets

**_My ask is is there any inbuilt or some customisations we can write to do filtering via bitsets in milvus ?_**

I saw this Enhancement request https://github.com/milvus-io/milvus/pull/32902 can this be somehow helpful ?

Thanks for the help

**Reply by xiaofan-luan** [reply]
but how do bitsets map to lucuene id? and document describe the feature?

**Reply by kiran12-lgtm** [reply]
BitDocSet represents an unordered set of Lucene Document Ids using a BitSet. A set bit represents inclusion in the set for that document. [Link](https://solr.apache.org/docs/6_0_1/solr-core/org/apache/solr/search/BitDocSet.html)

You can extend a solr Filter to provide bitset and use it wth RoaringDocIdSet to filter results faster
